### ISSUE TITLE
New User Approve is bypassing registration_errors hook

### ISSUE THREAD

**Issue by timothyadamlewis** [picklewagon/new-user-approve#76]
I discovered what *may be* a bug in new_user_approve.php, in the create_new_user function.  I notice you added code to check for the errors ( $errors->get_error_code() ), but I think you also need apply the registration_errors filter prior to this call:

```
$errors = apply_filters( 'registration_errors', $errors, $sanitized_user_login, $user_email );

                if ( $errors->get_error_code() ) {
                        return;
                }
```

The reason is that, in wp-includes/user.php, the call to apply the registration_errors filter appears AFTER the call to execute the register_post action (this is the action you hook to call your create_new_user function).  As a result, registration_errors doesn't get called, and I've found several plugins use it during the user registration (e.g., the CM Email Blacklist plugin).  This has caused the affected plugins to be bypassed when using New User Approve.  I have made the change to my local version, and it works fine now.

**Comment by d2roth** [comment]
@timothyadamlewis I was just about to report the same bug.

I am having similar issues to what you explain and so I am digging more into New User Approve to see why they do what they do.

From what I can see it looks like this is the chain of events:
1. `do_action( 'register_post', ... )` in wp-includes/user.php register_new_user();
2. `add_action( 'register_post', array( $this, 'create_new_user' ), 10, 3 );` in new-user-approve/new-user-approve.php __construct();
 - This now creates the new user
3. `$errors = apply_filters( 'registration_errors', ... );` in wp-includes/user.php register_new_user();
 - The user is available at this point now
4. `add_filter( 'registration_errors', array( $this, 'show_user_pending_message' ) );` is run in new-user-approve/new-user-approve.php __construct();
 - If there is a redirect_to destination or errors they are honoured but the user is already created. back in step 2 by create_new_user.
5. Other plugins run their filters on registration_errors without knowing the user is already created.

To solve this I believe New User Approve would need to change their logic so that they use `register_new_user` to add meta-data to the user and then check for that in registration_errors. I believe the logic would need to be changed like this:
1. `add_action( 'register_post', array( $this, 'create_new_user' ), 10, 3 )` would be removed
2. `add_filter( 'registration_errors', array( $this, 'show_user_pending_message' ) );` would be made a higher priority so that it runs after any other errors.
3. `show_user_pending_message` would get a check if the user is pending or not.

I think the reason for the way the plugin is programmed currently is for user experience. I am going to have to play around with it since I would like New User Approve to work with other plugins that use registration_errors to fail registration.

If I come up with anything I might just post back here my conclusions.

**Comment by timothyadamlewis** [comment]
Hi there!
I read your post, and what you say makes sense, but I'm not sure about the approach you propose.  What I do know is that I had plugins that appeared to be ineffective, such as a whitelisting plugin that blocks user accounts not listed in my whitelist.  I eventually traced it to the registration_errors firing after user creation - a big no-no.  So, I made the change I suggested to my own system, and that is what fixed it for me.  In my mind, the developer did what he had to do to get the functionality he desired, but it ended up subverting those plugins that depend on registration_errors.

Have you tried my suggestion to see if it solves your problem?

Thanks,
Tim

**Comment by d2roth** [comment]
Hey @timothyadamlewis,
No I have not tried your solution yet. The issue I see with it though is that some filters may be run twice. Anything with a priority less than the default 10 will be run twice actually. First before the new-user-approve call then inside it.

I spent time today reorganizing the plugin but I still need to test that its original purpose is kept.

If I can get it to work would you be willing to give it a test to see whether things operate properly on your end with my changes?

**Comment by d2roth** [comment]
I just found this pull request with a similar solution to mine: https://github.com/picklewagon/new-user-approve/pull/38

**Comment by timothyadamlewis** [comment]
I wondered about that - the impact of running registration_errors twice.
Didn't seem to be a problem, especially since the second call - the one
after register_post in user.php - is not used.

Sure - you try mine, and I'll try yours!

**Comment by d2roth** [comment]
@timothyadamlewis you can see what I have here: https://github.com/d2roth/new-user-approve/tree/develop

As I was finishing up I realized that any code hooked into the `register_new_user` action will not be fired either. The only time I found that action hooked to in WordPress' core was when WordPress sends out the new user email. I am thinking that there might be other plugins that break besides just simple registration plugins because of using the registration_errors filter for doing work.

I will test your change shortly.

**Comment by d2roth** [comment]
I have just tested with your fix and this is what I get when registering:
![image](https://user-images.githubusercontent.com/17486071/53646343-f176a400-3c08-11e9-83a7-a1a1a9f5dc37.png)

If you do want to use new-user-approve to create the user I think the easiest option would actually to just be changing the priority of the create_new_user call when registering the action. Something like this:
```PHP
add_action( 'register_post', array( $this, 'create_new_user' ), 10, 3 );
```
would become
```PHP
add_action( 'register_post', array( $this, 'create_new_user' ), 9999, 3 );
```

**Comment by timothyadamlewis** [comment]
I thought the main problem was that new_user_approve is using register_post to create the user before user.php does a few lines later.  I am not clear on how the priority change would affect that.

**Comment by d2roth** [comment]
You are right, the problem is in the use of a filter to do what a hook should be doing.

The priority change means that the action will be run after most filters added to register_post. This is *not* ideal but it does require less testing.

Does that make sense?

**Comment by timothyadamlewis** [comment]
I see what you mean - you are DE-prioritizing new_user_approve to execute last in the register_post filter call.  The one thing I'm not clear on is how frequently register_post is used by other plugins.  My thinking was that registration_errors was more frequently used, and no amount of prioritizing of the register_post call will help you with making sure plugins using registration_errors will get processed before new_user_approve's user creation function gets executed.  That is, unless you make sure registration_errors is called before the new_user_approve user creation executes.

**Comment by timothyadamlewis** [comment]
I just checked my WP installation: only new_user_approve uses the register_post hook, but four of my plugins use registration_errors.

**Comment by d2roth** [comment]
So you are correct in that increasing the priority on the register_post call won't do much if everything uses registration_errors. 

That priority change only works when applied with your modification in the opening issue where you run registration_errors from inside registration_post. It works because essentially the plugin would be flipping the order of those two actions.

It is convoluted I will say.

**Comment by timothyadamlewis** [comment]
Indeed, it is convoluted.  At least in my case, the problem was simple: I needed new_user_approve to see the *real* content of the $errors array before it executes its user creation function so that it would fail properly based on one or another plugin's results.  Executing registration_errors at the front of the create_new_user function (the one hooked to register_post) seemed like the fix.

One other approach I tried was modifying user.php to put the registration_errors call before register_post.  Big no-no, of course.

**Comment by d2roth** [comment]
The issue is that user.php says, regarding user_errors "If any errors are present in $errors, this will abort the user's registration." So theoretically, no matter what position in register_errors you hook into you should never have a user created.

**Comment by timothyadamlewis** [comment]
Right - that is the behavior of user.php.  But new_user_approve bypasses that algorithm by hooking register_post, which occurs before user.php calls registration_errors.  That's why I assumed just making sure registration_errors was called first thing in new_user_approve's create_new_user should be the solution.

BTW: I am more of a hacker of PHP than a Wordpress developer, so apologies in advance if I am "short on the concept" on any of this stuff.

**Comment by d2roth** [comment]
No worries. 

It would be nice if the plugin author would chime in here with their thoughts.

**Comment by d2roth** [comment]
So I came across this bug again because we did a plugin update so I figured I needed to find a way to apply these fixes while allowing automatic updates. I now have built a mu-plugin that applies the required fixes to New User Approve without modifying the plugin files in any way.

You can see it here: https://gist.github.com/d2roth/d3373ff17b1e14cc651f6abfa4ebb1c3

Essentially I just deregistered their create_new_user function and assigned the rest of the methods to more appropriate hooks.

I hope this can be properly fixed by the plugin author though.

**Comment by timothyadamlewis** [comment]
Hi there!

Very kind of you to come up with a solution that supports automatic
updates!  I’ll give it a look.
